// montage/examples/feed-reader/main.reel/main.js
/* <copyright>
 This file contains proprietary software owned by Motorola Mobility, Inc.<br/>
 No rights, expressed or implied, whatsoever to this software are provided by Motorola Mobility, Inc. hereunder.<br/>
 (c) Copyright 2011 Motorola Mobility, Inc.  All Rights Reserved.
 </copyright> */
var Montage = require("montage/core/core").Montage,
    Component = require("montage/ui/component").Component;

exports.Main = Montage.create(Component, {

    feedURL: {
        value: 'http://rss.cnn.com/rss/cnn_mostpopular.rss',
        distinct: true
    },
    
    mapCategory: {
        get: function() {
            return this._mapCategory;
        },
        set: function(value) {
            this._mapCategory = value;
            this.map.category = value;
        }
    },
    
    feedReader: {value: null},
    
    map: {value: null},
    
    mapCenter: {value: null},
    
    handleGoAction: {
        value: function(event) {
            event.preventDefault();
            this.feedReader.feedURL = this.feedURL;
            
        }
    },
    
    handleMapButtonAction: {
        value: function(event) {
            event.preventDefault();            
            this.map.center = this.mapCenter;            
        }
    },
    
    draw: {
        value: function() {
            if(this.mapCenter != null) {
                this.map.center = this.mapCenter;
            }
        }
    }
    
});
